File: Exercicio_paradigmas.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define QtdIteracoes 25
#define TAM 30000000 // 30 million

void preencherVetor(int* v, int lenght) {
    int i;
    for(i=0;i<lenght;i++) {
        v[i] = rand()%lenght;
    }
}

// recebe uma cópia do valor dado no parametro e o retorna
int valor_formal(int valor) {
    return valor;
}

// a passagem ocorre por valor, logo os valores passados são os formais
void passagem_valor(int* vetorOrigem, int* vetorDestino, int lenght) {
    int i;
    for(i=0;i<lenght;i++) {
        vetorDestino[i] = valor_formal(vetorOrigem[i]);
    }
}

// a passagem ocorre por referencia, logo os valores passados são os reais
void passagem_referencia(int* vetorOrigem, int* vetorDestino, int lenght) {
    int i;
    for(i=0;i<lenght;i++) {
        vetorDestino[i] = vetorOrigem[i];
    }
}

int main() {
    srand(time(NULL));
	int i=1, j;
	
    int* vetorOrigem = (int*) malloc(sizeof(int)*TAM);
    int* vetorDestino = (int*) malloc(sizeof(int)*TAM);
    clock_t start, end;
    double total_t = 0;
	
	
	FILE* arq1 = fopen("valor.out","w");
	FILE* arq2 = fopen("referencia.out","w");
	
    while(i <= QtdIteracoes) {
    	printf("----------------- %i ------------------------", i);
    	preencherVetor(vetorOrigem, TAM);
    	// passagem por valor
    	start = clock();
	    passagem_valor(vetorOrigem, vetorDestino, TAM);
	    end = clock();
	    total_t = ((double) (end - start)) / CLOCKS_PER_SEC;
	    printf("\npassagem por valor:      %.3f segundos\n", total_t);
		
		j = total_t*1000;
		fprintf(arq1, "0,%i\n", j);
		
		// passagem por referência
	    start = clock();
	    passagem_referencia(vetorOrigem, vetorDestino, TAM);
	    end = clock();
	    total_t = ((double) (end - start)) / CLOCKS_PER_SEC;
	    printf("passagem por referencia: %.3f segundos\n", total_t);
	    
	    j = total_t*1000;
	    fprintf(arq2, "0,%i\n", j);
	    i++;
	}

	fclose(arq1);
	fclose(arq2);
}

File: Codigos_java/Exercicio_paradigmas/src/ExercicioParadigmas.java
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Random;

public class ExercicioParadigmas {
	public static final int QtdIteracoes = 25;
	public static final int TAM = 30000000; // 30 million
	public static Random rand = new Random();
	
	public static void preencherVetor(int[] v) {
		for(int i = 0; i<v.length ; i++) {
			v[i] = rand.nextInt(TAM);
		}
	}
	
	public static int valor_formal(int valorFormal) {
		return valorFormal;
	}
	
	public static void passagem_valor(int[] vetorOrigem, int[] vetorDestino) {
		for(int i = 0 ; i<vetorOrigem.length ; i++) {
			vetorDestino[i] = valor_formal(vetorOrigem[i]);
		}
	}
	
	public static void passagem_referencia(int[] vetorOrigem, int[] vetorDestino) {
		for(int i = 0 ; i<vetorOrigem.length ; i++) {
			vetorDestino[i] = vetorOrigem[i];
		}
	}
	
	public static boolean equals(int[] v1, int[] v2) {
		for(int i=0; i< v1.length ; i++ ) {
			if(v1[i] != v2[i]) {
				return false;
			}
		}
		return true;
			
	}
	
	public static void main(String[] args) throws IOException{
		int[] vetorOrigem = new int[TAM];
		int[] vetorDestino = new int[TAM];
		long start, end;
		double total;
		
		int i=1;
		//preencherVetor(vetorOrigem);
		
		FileWriter arq1 = new FileWriter("D:\\Pepê\\2018.1 UFRPE\\Paradigmas de Programação\\Exercicio_2\\referencia2.out");
	    PrintWriter gravarArq1 = new PrintWriter(arq1);
	    FileWriter arq2 = new FileWriter("D:\\Pepê\\2018.1 UFRPE\\Paradigmas de Programação\\Exercicio_2\\valor2.out");
	    PrintWriter gravarArq2 = new PrintWriter(arq2);
	    	
		while(i <= QtdIteracoes) {
			System.out.printf("------------------ %d ----------------------", i);
			preencherVetor(vetorOrigem);
			start = System.currentTimeMillis();
			passagem_valor(vetorOrigem, vetorDestino);
			end = System.currentTimeMillis();
			total = (double) (end - start)/1000;
			System.out.printf("\npassagem por valor:      %.3f segundos\n", total);
			
			gravarArq1.printf("%.3f\n",total);
			
			start = System.currentTimeMillis();
			passagem_referencia(vetorOrigem, vetorDestino);
			end = System.currentTimeMillis();
			total = (double) (end - start)/1000;
			System.out.printf("passagem por referencia: %.3f segundos\n", total);
			
			gravarArq2.printf("%.3f\n",total);
			 i++;
		}
				
	    arq1.close();
	    arq2.close();
	}
}

